File: entities/enums.py
# entities/enums.py
"""
标签系统枚举模块
主要功能：统一定义游戏实体所需的分类标签，使用标准的枚举类替换硬编码字符串，遵循开闭原则。
"""
from enum import Enum

class FactionTag(str, Enum):
    """
    完整阵营枚举类
    主要功能：定义目前系统支持的15个阵营及预留的扩展项。
    """
    GENTLE_HOUSE = "Gentle_House"                # 狡兔屋
    INVESTIGATION_UNIT = "Criminal_Investigation" # 刑侦特勤组
    HSOS6 = "H.S.O.S.6"                          # H.S.O.S.6
    BELOBOG_HEAVY_IND = "Belobog_Heavy_Ind"      # 白祇重工
    VICTORIA_HOUSEKEEPING = "Victoria_Housekeeping" # 维多利亚家政
    SONS_OF_CALYDON = "Sons_of_Calydon"          # 卡吕冬之子
    LYRA = "Lyra"                                # 天琴座
    OBOLUS_SQUAD = "Obolus_Squad"                # 奥波勒斯小队
    DEFENSE_FORCE_SILVER = "Defense_Force_Silver" # 防卫军·白银小队
    MOCKINGBIRD = "Mockingbird"                  # 反舌鸟
    YUN_KUI_MOUNTAIN = "Yun_Kui_Mountain"        # 云岿山
    ODD_EATER_HOUSE = "Odd_Eater_House"          # 怪啖屋
    KRAMPUS_DARK_BRANCH = "Krampus_Dark_Branch"  # 坎卜斯黑枝
    DELUSIONAL_ANGEL = "Delusional_Angel"        # 妄想天使
    CITY_ORDER_DEPT = "City_Order_Dept"          # 都市秩序部
    OTHER = "Other"                              # 其他/未来新增

class SpecialtyTag(str, Enum):
    """
    角色特性枚举类
    主要功能：定义角色在战斗体系中的定位特性。
    """
    ATTACK = "Attack"     # 强攻
    STUN = "Stun"         # 击破
    SUPPORT = "Support"   # 支援
    RUPTURE = "Rupture"   # 命破
    ANOMALY = "Anomaly"   # 异常
    DEFENSE = "Defense"   # 防护

class ElementTag(str, Enum):
    """
    元素属性枚举类
    主要功能：定义角色及异常状态对应的元素伤害类型。
    """
    ICE = "Ice"
    FIRE = "Fire"
    ETHER = "Ether"
    PHYSICAL = "Physical"
    ELECTRIC = "Electric"
    WIND = "Wind"
    AURIC_INK = "Auric Ink"     # 特殊以太-玄墨
    FROST = "Frost"             # 特殊冰-烈霜
    HONED_EDGE = "Honed Edge"   # 特殊物理-凛刃

class CharacterState(str, Enum):
    """
    角色驻场状态枚举类
    主要功能：标识角色当前处于前台、后台或不可交互状态。
    """
    ACTIVE = "Active"       # 前台
    STANDBY = "Standby"     # 后台
    LOCKED = "Locked"       # 不可交互

class EnemyType(str, Enum):
    """
    敌人类型枚举类
    主要功能：区分不同量级的敌人，后续用于限制连携技次数等逻辑。
    """
    NORMAL = "Normal"       # 连携技1次
    ELITE = "Elite"         # 连携技2次
    BOSS = "Boss"           # 连携技3次

File: zsim-core/src/entities/enums.rs
use serde::{Deserialize, Serialize};

/// 标识角色所属派系的阵营标签。
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum FactionTag {
    #[serde(rename = "Gentle_House")]
    GentleHouse,
    #[serde(rename = "Criminal_Investigation")]
    CriminalInvestigation,
    #[serde(rename = "H.S.O.S.6")]
    Hsos6,
    #[serde(rename = "Belobog_Heavy_Ind")]
    BelobogHeavyInd,
    #[serde(rename = "Victoria_Housekeeping")]
    VictoriaHousekeeping,
    #[serde(rename = "Sons_of_Calydon")]
    SonsOfCalydon,
    #[serde(rename = "Lyra")]
    Lyra,
    #[serde(rename = "Obolus_Squad")]
    ObolusSquad,
    #[serde(rename = "Defense_Force_Silver")]
    DefenseForceSilver,
    #[serde(rename = "Mockingbird")]
    Mockingbird,
    #[serde(rename = "Yun_Kui_Mountain")]
    YunKuiMountain,
    #[serde(rename = "Odd_Eater_House")]
    OddEaterHouse,
    #[serde(rename = "Krampus_Dark_Branch")]
    KrampusDarkBranch,
    #[serde(rename = "Delusional_Angel")]
    DelusionalAngel,
    #[serde(rename = "City_Order_Dept")]
    CityOrderDept,
    #[serde(rename = "Other")]
    Other,
}

/// 标识角色战斗定位的特化标签。
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SpecialtyTag {
    #[serde(rename = "Attack")]
    Attack,
    #[serde(rename = "Stun")]
    Stun,
    #[serde(rename = "Support")]
    Support,
    #[serde(rename = "Rupture")]
    Rupture,
    #[serde(rename = "Anomaly")]
    Anomaly,
    #[serde(rename = "Defense")]
    Defense,
}

/// 用于伤害类型和异常交互的元素标签。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ElementTag {
    #[serde(rename = "Ice")]
    Ice,
    #[serde(rename = "Fire")]
    Fire,
    #[serde(rename = "Ether")]
    Ether,
    #[serde(rename = "Physical")]
    Physical,
    #[serde(rename = "Electric")]
    Electric,
    #[serde(rename = "Auric Ink")]
    AuricInk,
    #[serde(rename = "Frost")]
    Frost,
    #[serde(rename = "Honed Edge")]
    HonedEdge,
}

/// 角色在场状态。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
pub enum CharacterState {
    #[serde(rename = "Active")]
    Active,
    #[serde(rename = "Standby")]
    #[default]
    Standby,
    #[serde(rename = "Locked")]
    Locked,
}

/// 敌人等级分类，用于连携攻击次数限制。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
pub enum EnemyType {
    #[serde(rename = "Normal")]
    #[default]
    Normal,
    #[serde(rename = "Elite")]
    Elite,
    #[serde(rename = "Boss")]
    Boss,
}

impl EnemyType {
    pub fn chain_attack_limit(&self) -> u32 {
        match self {
            EnemyType::Normal => 1,
            EnemyType::Elite => 2,
            EnemyType::Boss => 3,
        }
    }
}

/// 角色技能动作的类别。
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum SkillType {
    #[serde(rename = "Normal")]
    Normal,
    #[serde(rename = "Special")]
    Special,
    #[serde(rename = "Ultimate")]
    Ultimate,
    #[serde(rename = "Chain")]
    Chain,
    #[serde(rename = "Dodge")]
    Dodge,
    #[serde(rename = "Assist")]
    Assist,
    #[serde(rename = "Coordinated")]
    Coordinated,
    #[serde(rename = "QuickAssist")]
    QuickAssist,
}

/// 连携攻击和反应技能的触发条件。
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum TriggerType {
    #[serde(rename = "OnDamageDealt")]
    OnDamageDealt,
    #[serde(rename = "OnAnomalyTriggered")]
    OnAnomalyTriggered,
    #[serde(rename = "OnChainAttack")]
    OnChainAttack,
    #[serde(rename = "OnDodge")]
    OnDodge,
    #[serde(rename = "OnParry")]
    OnParry,
}

/// 模拟执行模式。
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum SimMode {
    #[serde(rename = "Single")]
    Single,
    #[serde(rename = "Parallel")]
    Parallel,
    #[serde(rename = "Full")]
    Full,
    #[serde(rename = "Loop")]
    Loop,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_faction_serde_roundtrip() {
        let factions = vec![
            FactionTag::GentleHouse,
            FactionTag::CriminalInvestigation,
            FactionTag::Other,
        ];
        for f in factions {
            let json = serde_json::to_string(&f).expect("serialize");
            let back: FactionTag = serde_json::from_str(&json).expect("deserialize");
            assert_eq!(f, back);
        }
    }

    #[test]
    fn test_faction_deserialize_from_python() {
        let json = r#""Gentle_House""#;
        let f: FactionTag = serde_json::from_str(json).expect("deserialize");
        assert_eq!(f, FactionTag::GentleHouse);
    }

    #[test]
    fn test_specialty_serde() {
        let json = r#""Attack""#;
        let s: SpecialtyTag = serde_json::from_str(json).expect("deserialize");
        assert_eq!(s, SpecialtyTag::Attack);
    }

    #[test]
    fn test_element_serde() {
        let json = r#""Fire""#;
        let e: ElementTag = serde_json::from_str(json).expect("deserialize");
        assert_eq!(e, ElementTag::Fire);
    }

    #[test]
    fn test_character_state_serde() {
        let json = r#""Active""#;
        let s: CharacterState = serde_json::from_str(json).expect("deserialize");
        assert_eq!(s, CharacterState::Active);
    }

    #[test]
    fn test_enemy_type_chain_limits() {
        assert_eq!(EnemyType::Normal.chain_attack_limit(), 1);
        assert_eq!(EnemyType::Elite.chain_attack_limit(), 2);
        assert_eq!(EnemyType::Boss.chain_attack_limit(), 3);
    }

    #[test]
    fn test_enemy_type_serde() {
        let json = r#""Boss""#;
        let et: EnemyType = serde_json::from_str(json).expect("deserialize");
        assert_eq!(et, EnemyType::Boss);
    }

    #[test]
    fn test_sim_mode_serde() {
        let json = r#""Parallel""#;
        let m: SimMode = serde_json::from_str(json).expect("deserialize");
        assert_eq!(m, SimMode::Parallel);
    }
}
